ceph/ceph_admin/__init__.py

```diff
@@ -0,0 +1,764 @@
+            config: test data configuration
+
+        config:
+            base_url: ceph compose URL
+            container_image: custom ceph container image
+        """
+        self.cluster = cluster
+        self.config = config
+        self.installer = self.cluster.get_ceph_object("installer")
+
+    def read_cephadm_gen_pub_key(self):
+        """
+        Read cephadm generated public key
+        Arg:
+            Installer node
+        Returns:
+            Public Key string
+        """
+        ceph_pub_key, _ = self.installer.exec_command(
+            sudo=True, cmd="cat /etc/ceph/ceph.pub"
+        )
+        return ceph_pub_key.read().decode().strip()
+
+    def distribute_cephadm_gen_pub_key(self, nodes=None):
+        """
+        Distribute cephadm generated public key to all nodes in the list.
+        Args:
+            nodes: node list to add ceph public key(default: None)
+        """
+        ceph_pub_key = self.read_cephadm_gen_pub_key()
+
+        if nodes is None:
+            nodes = self.cluster.get_nodes()
+
+        nodes = nodes if isinstance(nodes, list) else [nodes]
+
+        for each_node in nodes:
+            each_node.exec_command(sudo=True, cmd="install -d -m 0700 /root/.ssh")
+            keys_file = each_node.write_file(
+                sudo=True, file_name="/root/.ssh/authorized_keys", file_mode="a"
+            )
+            keys_file.write(ceph_pub_key)
+            keys_file.flush()
+            each_node.exec_command(
+                sudo=True, cmd="chmod 0600 /root/.ssh/authorized_keys"
+            )
+
+    def check_exist(self, daemon, ids, timeout=None, interval=5):
+        """
+        Validate daemon existence using daemon ids
+
+        this method uses `ceph orch ls -f json`
+        to get JSON formatted output and
+        validates provided daemon ids with timeout.
+
+        Args:
+            daemon: daemon name
+            ids: daemon ids
+            timeout: timeout in seconds
+            interval: interval in seconds
+
+        Returns:
+            boolean
+        """
+        timeout = self.TIMEOUT if not timeout else timeout
+        checks = timeout / interval
+
+        while checks:
+            checks -= 1
+
+            out, _ = self.shell(
+                remote=self.installer,
+                args=["ceph", "orch", "ps", "-f", "json-pretty"],
+            )
+
+            out = json.loads(out)
+            daemons = dict(
+                (i["daemon_id"], i["status_desc"])
+                for i in out
+                if i.get("daemon_type") == daemon
+            )
+
+            count = 0
+            for _id in ids:
+                if _id in daemons:
+                    if daemons[_id] == "running":
+                        count += 1
+
+            logger.info(
+                "%s/%s %s daemon(s) up... Retries: %s"
+                % (count, len(ids), daemon, checks)
+            )
+            if count == len(ids):
```

Comment:
Would it make sense moving this statement above the log statement?

Reply:
Yes we can, but we may need log statement to say "all daemons are running" before returning True value, that means progressively we may miss "2/2 daemons up and running" before returning.

or may be we can add logger statement to log "all daemons are up and running" before returning.

I also think that currently both(progress and final statuses) dumped in current logging with single log expression by not moving after check `count == len(ids)`.

Reply:
Okay, I agree with your assessment.